nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -63,9 +64,85 @@ impl<'a> Planner<'a> {
     }
 
     fn do_plan(&mut self) -> Result<(), Error> {
+        // XXX fix below comment
         // The only thing this planner currently knows how to do is add services
         // to a sled that's missing them.  So let's see if we're in that case.
 
+        self.do_plan_expunge()?;
+        self.do_plan_add()
+    }
+
+    fn do_plan_expunge(&mut self) -> Result<(), Error> {
+        fn expunge_reason(
+            log: &Logger,
+            sled_info: &SledResources,
+        ) -> Option<&'static str> {
+            // We could in principle have a decommissioned sled which has
+            // resources allocated to it -- if we do, then we should log this
+            // as an error because that is not a legal state, but remove those
+            // zones anyway.
+            match sled_info.state {
+                SledState::Active => {}
+                SledState::Decommissioned => {
+                    error!(
+                        &log,
+                        "sled has state Decommissioned, yet has resources allocated to it; \
+                         removing them here (sled policy is \"{}\")", sled_info.policy;
+                    );
+
+                    return Some("decommissioned");
+                }
+            }
+
+            match sled_info.policy {
+                SledPolicy::InService { .. } => None,
+                SledPolicy::Expunged => Some("expunged"),
+            }
+        }
```

Comment:
Is it possible to have a sled in state "decommissioned" that's not got policy "expunged"?  (If not, I think a lot of this could be simplified -- there'd only be one possible expunge reason and you'd only need to check one field to identify it.)